src/java.base/share/classes/com/sun/crypto/provider/GCTR.java

```diff
@@ -112,9 +116,87 @@ int update(byte[] in, int inOfs, int inLen, byte[] out, int outOfs) {
         }
     }
 
+    // input must be multiples of AES blocks, 128-bit, when calling update
+    int update(byte[] in, int inOfs, int inLen, ByteBuffer dst) {
+        if (inLen - inOfs > in.length) {
+            throw new RuntimeException("input length out of bound");
+        }
+        if (inLen < 0 || inLen % AES_BLOCK_SIZE != 0) {
+            throw new RuntimeException("input length unsupported");
+        }
+        if (dst.remaining() < inLen) {
+            throw new RuntimeException("output buffer too small");
+        }
+
+        long blocksLeft = blocksUntilRollover();
+        int numOfCompleteBlocks = inLen / AES_BLOCK_SIZE;
+        if (numOfCompleteBlocks >= blocksLeft) {
+            // Counter Mode encryption cannot be used because counter will
+            // roll over incorrectly. Use GCM-specific code instead.
+            byte[] encryptedCntr = new byte[AES_BLOCK_SIZE];
+            for (int i = 0; i < numOfCompleteBlocks; i++) {
+                embeddedCipher.encryptBlock(counter, 0, encryptedCntr, 0);
+                for (int n = 0; n < AES_BLOCK_SIZE; n++) {
+                    int index = (i * AES_BLOCK_SIZE + n);
+                    dst.put((byte) ((in[inOfs + index] ^ encryptedCntr[n])));
+                }
+                GaloisCounterMode.increment32(counter);
+            }
+            return inLen;
+        } else {
+            int len = inLen - inLen % AES_BLOCK_SIZE;
+            int processed = len;
+            byte[] out = new byte[Math.min(MAX_LEN, len)];
+            int offset = inOfs;
+            while (processed > MAX_LEN) {
+                encrypt(in, offset, MAX_LEN, out, 0);
+                dst.get(out, 0, MAX_LEN);
```

Comment:
Shouldn't this be "put" instead of "get"?

Reply:
Yeah.. I'm surprised that wasn't caught by the tests.  I will look to see what case I need to make to check that.

Reply:
So there is only one calling method that never gives this method it more than 16 bytes.  That is why testing never caught this.  But it doesn't hurt performance leaving the ability to call larger input sizes.